go/types/graph_builder.go

```diff
@@ -0,0 +1,263 @@
+// Copyright 2016 Attic Labs, Inc. All rights reserved.
+// Licensed under the Apache License, version 2.0:
+// http://www.apache.org/licenses/LICENSE-2.0
+
+package types
+
+import (
+	"bytes"
+	"fmt"
+
+	"github.com/attic-labs/noms/go/d"
+	"github.com/attic-labs/noms/go/util/status"
+	"github.com/dustin/go-humanize"
+)
+
+type GraphBuilder struct {
+	oc      opCache
+	vrw     ValueReadWriter
+	stack   graphStack
+	verbose bool
+}
+
+// NewGraphBuilder() returns an object that is used in building a graph of
+// objects that are stored on disk and then assembled into a noms collection.
+// Graphs can exist of nested maps, with maps, sets and lists at the outermost
+// levels of the graph. Any of the collections (nested or leaf) can contain onto
+// any type of Value. |rootKind| indicates what the topLevel collection should be.
+// It must be one of: MapKind, SetKind, or ListKind. If this GraphBuilder is
+// intended to build nested maps, |rootKind| must be MapKind.
+func NewGraphBuilder(vrw ValueReadWriter, rootKind NomsKind, verbose bool) *GraphBuilder {
+	return newGraphBuilder(vrw, vrw.opCache(), rootKind, verbose)
+}
+
+func newGraphBuilder(vrw ValueReadWriter, opc opCache, rootKind NomsKind, verbose bool) *GraphBuilder {
+	b := &GraphBuilder{oc: opc, vrw: vrw, verbose: verbose}
+	b.pushNewKeyOnStack(String("ROOT"), rootKind, nil)
+	return b
+}
+
+// MapSet() will add the key/value pair |k, v| to the map found by traversing
+// the graph using the |keys| slice. Intermediate maps referenced by |keys| are
+// created as necessary
+func (b *GraphBuilder) MapSet(keys []Value, k Value, v Value) {
+	d.Chk.True(b.oc != nil, "Can't call MapSet() again after Build()")
+	b.oc.GraphMapSet(keys, k, v)
+}
+
+// SetInsert() will insert the value |v| into the set at path |keys|. Intermediate
+// maps referenced by |keys| are created as necessary.
+func (b *GraphBuilder) SetInsert(keys []Value, v Value) {
+	d.Chk.True(b.oc != nil, "Can't call SetInsert() again after Build()")
+	b.oc.GraphSetInsert(keys, v)
+}
+
+// ListAppends() will append |v| to the list at path |p|. Intermediate
+// maps referenced by |keys| are created as necessary.
+func (b *GraphBuilder) ListAppend(keys []Value, v Value) {
+	d.Chk.True(b.oc != nil, "Can't call ListAppend() again after Build()")
+	b.oc.GraphListAppend(keys, v)
+}
+
+type GraphOpContainer struct {
+	keys []Value
+	kind NomsKind
+	item sequenceItem
+}
+
+// Builds and returns the graph.
+func (b *GraphBuilder) Build() Value {
+	d.Chk.True(b.oc != nil, "Can only call Finish() once")
```

Comment:
I'm not really asserting an opinion about whether this thing should be goroutine safe right now, just that you should make a decision, document it and have it the provided semantics be correct.

in the abstract, having this threadsafe seems highly useful because the basic goal here is to be able to efficient operate over data which won't fit in memory. it seems likely we're going to want to do things like that in parallel.

Abstractly, it would make more sense to me that `ListAppend` because `ListSet(index uint64, v Value)`. If you do that, then you get a couple of things:

 - It's more straightforward how to make this threadsafe
 - You have an abstraction which allows you specify deterministically an location in graph.
 - Lists could then be internal nodes in trees, which seems somewhat likely to be useful (i.e. list of structs).
 - The mental model of how GraphBuilder works is self consistent (I add items to a graph based on path to their location, in any order => I can create a graph of any size).

I'm fine with proceeding in multiple directions. For instance, documenting this as non-threadsafe and landing as is, is an acceptable answer -- leaving the more general solution for later.